File: 135.classes_traits_objects/src/main/scala/solution/ClassesTraitsObjects.scala
package solution;

import collection.immutable._
import scala.collection.mutable.ListBuffer

object Garage {
  val vehicles = new ListBuffer[Vehicle]()

  def addVehicle(vehicle: Vehicle) =  { vehicles += vehicle; this }
  def removeVehicle(vehicle: Vehicle) = { vehicles -= vehicle; this }
  def getVehicles() : List[Vehicle] = vehicles.toList
}

case class Vehicle(make: String, wheelCount : Int)
class Car(make: String) extends Vehicle(make, 4) 
class Truck(make: String) extends Vehicle(make, 18) 

trait Convertible {
  def openTop() { println("Top is open") }
  def closeTop() { println("Top is closed") }
}

class SportsCar(make: String) extends Vehicle(make, 4) with Convertible

object ClassesTraitsObjects {
  def main(args: Array[String]): Unit = {
    val garage = Garage
    garage.addVehicle(new Car("BMW"))
    garage.addVehicle(new SportsCar("Maserati"))
    garage.addVehicle(new Truck("Scania"))

    garage.getVehicles foreach {
      v => println("This vehicle is a " + v.make + " and has " + v.wheelCount + " wheels");

      v match {
        case c:Convertible => {
                  println("Oooh, this vehicle is a convertible. Let's play with the top!")
                  c.openTop()
                  c.closeTop()
        }
        case _ =>
      }
    }
  }
}

File: 135.classes_traits_objects/src/main/java/javax/ClassesTraitsObjects.java
package javax;

import java.util.*;

enum Garage {
  INSTANCE;
  private final static List<Vehicle> vehicleList = new LinkedList<Vehicle>();

  public static void addVehicle(Vehicle vehicle) {
    vehicleList.add(vehicle);
  }

  public static List<Vehicle> getVehicles() {
    return vehicleList;
  }
}

class Vehicle {
  private final String make;
  private final int wheelCount;

  protected Vehicle(final String make, final int wheelCount) { 
    this.make = make; 
    this.wheelCount = wheelCount; 
  }

  protected String getMake() { return make; }
  protected int getWheelCount() { return wheelCount; }
}

class Car extends Vehicle {
  public Car(final String make) { 
    super(make, 4); 
  }
}

class Truck extends Vehicle {
  public Truck(final String make) { 
    super(make, 18); 
  }
}

interface Convertible {
  void openTop();
  void closeTop();
}

class SportsCar extends Car implements Convertible {
  public SportsCar(final String make) { 
    super(make); 
  }

  @Override
  public void openTop() {
    System.out.println("Top is open");
  }

  @Override
  public void closeTop() {
    System.out.println("Top is closed");
  }
}

public class ClassesTraitsObjects {
  public static void main(String[] args) {
    Garage.addVehicle(new Car("BMW"));
    Garage.addVehicle(new SportsCar("Maserati"));
    Garage.addVehicle(new Truck("Scania"));

    for (Vehicle vehicle : Garage.getVehicles()) {
      System.out.println("This vehicle is a " + vehicle.getMake() + " and has " + vehicle.getWheelCount() + " wheels");
      if (vehicle instanceof Convertible) {
        System.out.println("Oooh, this vehicle is a convertible. Let's play with the top!");
        Convertible convertible = (Convertible)vehicle;
        convertible.openTop();
        convertible.closeTop();
      }
    }
  }
}
